File: solutions/code-carl/ch04_string/T5_LC0459_RepeatedSubstringPattern.java
/**
 * Practice template for:
 * - LeetCode 459. Repeated Substring Pattern
 *   https://leetcode.cn/problems/repeated-substring-pattern/
 */
public class T5_LC0459_RepeatedSubstringPattern {
    public boolean repeatedSubstringPattern(String s) {
        char[] chars = s.toCharArray();
        int[] next = new int[chars.length];
        int j = 0;
        for (int i = 1; i < chars.length; i++) {
            while (j > 0 && chars[i] != chars[j])
                j = next[j - 1];
            if (chars[i] == chars[j])
                j++;
            next[i] = j;
        }
        int len = s.length() - next[s.length() - 1];
        return len <= s.length() / 2 && s.length() % len == 0;
    }
}

File: archive/2026-09-18/ch04_string/t5_lc0459_repeated_substring_pattern.py
"""LeetCode 459. Repeated Substring Pattern
https://leetcode.cn/problems/repeated-substring-pattern/
"""
from typing import List


class Solution:
    def repeatedSubstringPattern(self, s: str) -> bool:
        j = 0
        next: List[int] = [0] * len(s)
        for i in range(1, len(s)):
            while j > 0 and s[i] != s[j]:
                j = next[j - 1]
            if s[i] == s[j]:
                j += 1
            next[i] = j

        return len(s) > next[-1] > 0 == len(s) % (len(s) - next[-1])


if __name__ == "__main__":
    solution = Solution()
    assert solution.repeatedSubstringPattern("abab") is True
    assert solution.repeatedSubstringPattern("aba") is False
    assert solution.repeatedSubstringPattern("abcabcabcabc") is True
    print("All tests passed.")
